src/BlockTypes/ProductQuery.php

```diff
@@ -66,28 +45,30 @@ public function get_query_by_attributes_once( $query ) {
 	 * @param array       $parsed_block The block being rendered.
 	 */
 	public function update_query( $pre_render, $parsed_block ) {
-		if ( 'core/query' !== $parsed_block['blockName'] || ! isset( $parsed_block['attrs']['__woocommerceVariationProps'] ) ) {
+		if ( 'core/query' !== $parsed_block['blockName'] ) {
 			return;
 		}
 
 		$this->parsed_block = $parsed_block;
 
-		add_filter(
-			'query_loop_block_query_vars',
-			array( $this, 'get_query_by_attributes_once' ),
-			10,
-			1
-		);
+		if ( isset( $parsed_block['attrs']['__woocommerceVariationProps'] ) ) {
+			add_filter(
+				'query_loop_block_query_vars',
+				array( $this, 'get_query_by_attributes' ),
```

Comment:
No, we shouldn't. 

> This way we don't need to set $this->parsed_block anymore.

It is still necessary this. If we don't use this anymore, the $parsed_block will always be an instance of our variations (for the check at line 54). This is a problem because, for example, if on the same page, we have:
- Products Query block.
- Query Loop block with post type set to `Post` (so it should render the post on frontend).

The filter injects in the Query Loop block the query used by the Products Query block. Obviously, this is incorrect.

Let me know what you think, and if it makes sense to you, I will merge this PR!

Reply:
Re-reading this code, I wonder… is there any way we can make these lines conditional?

https://github.com/woocommerce/woocommerce-blocks/pull/7109/files#diff-efb947e345af5e501dfaa2980b055b3b31c308b31e8db23049b6b48fecee33d8L31-L36

Like, is there any other filter we can use so that we don't run this on every block? It would be amazing if there ware a filter like `rest_{$this->post_type}_query`, but with something like `pre_render_{$this->block_id}_block`. You know what I mean? 🤔

Reply:
> Re-reading this code, I wonder… is there any way we can make these lines conditional?

GitHub doesn't highlight the lines. From the URL, it seems that you are referring to these:
 
![image](https://user-images.githubusercontent.com/4463174/190344134-a251c2ba-0d9e-43c2-b18f-f8313b4ff3cb.png)

If yes, we already checked during one of our pair sessions: these lines are executed once: when the class is instanced.

> Like, is there any other filter we can use so that we don't run this on every block? It would be amazing if there ware a filter like rest_{$this->post_type}_query, but with something like pre_render_{$this->block_id}_block. You know what I mean? 🤔

Yeah, I searched during my exploration, but unfortunately doesn't exist.

Reply:
Oooh, right, thanks for the explanation. Another reason not to use an inline function is to avoid the same filter being added several times.

Reply:
> Another reason not to use an inline function is to avoid the same filter being added several times.

Not sure what you mean by `the same filter being added several times`. With this code, we add the filter for each variation that is shown on the page.

Reply:
I might have been confused. But with this PR, the filter isn't added for each block instance, that's because we are calling a class method so every time `add_filter()` is called, it knows it's the same method. But if the function was inline, it would be added as a filter every time it's called. (I might be wrong, though)

Reply:
> If yes, we already checked during one of our pair sessions: these lines are executed once: when the class has been instanced.

@gigitux  Sorry, I wasn't clear enough. I know these lines get executed once in the sense that the filter is registered only once. But the filter is run on the pre-render of every block. Of course we return early and no harm is done. But I still would have loved if it wasn't like that…

I hope this makes more sense.

Reply:
> I hope this makes more sense.

It makes sense! 

> Of course we return early and no harm is done. But I still would have loved if it wasn't like that…
Agree, but it seems that it isn't possible :( 

> I might have been confused. But with this PR, the filter isn't added for each block instance, that's because we are calling a class method so every time add_filter() is called, it knows it's the same method. But if the function was inline, it would be added as a filter every time it's called. (I might be wrong, though)

The filter `query_loop_block_query_vars` is added every time that the Product Query variation is rendered.
The reason is that `pre_render_block` is triggered on the pre-render of every block.

Does it make sense?

cc @sunyatasattva

Reply:
It makes sense! I did a test, and I can confirm what you wrote!
Just to be clear, in our case the function is called 4 times for each block. The reason is that the function `gutenberg_build_query_vars_from_query_block` (where there is the filter `query_loop_block_query_vars`) in the hierarchy of the Query Block is called 4 times:
- [post-template.php](https://github.com/WordPress/gutenberg/blob/8974d3b394964fc255b8ccf1c6a6b45c07eb4fe7/packages/block-library/src/post-template/index.php/#L53-L54)
- [query-no-results.php](https://github.com/WordPress/gutenberg/blob/8974d3b394964fc255b8ccf1c6a6b45c07eb4fe7/packages/block-library/src/query-no-results/index.php/#L31-L32)
- [query-pagination-next.php](https://github.com/WordPress/gutenberg/blob/8974d3b394964fc255b8ccf1c6a6b45c07eb4fe7/packages/block-library/src/query-pagination-next/index.php/#L47-L48)
- [query-pagination-numbers.php](https://github.com/WordPress/gutenberg/blob/8974d3b394964fc255b8ccf1c6a6b45c07eb4fe7/packages/block-library/src/query-pagination-numbers/index.php)